perun-core/src/main/java/cz/metacentrum/perun/core/impl/ExtSourceTCS.java

```diff
@@ -0,0 +1,311 @@
+		throw new ExtSourceUnsupportedOperationException();
+	}
+
+	@Override
+	public Map<String, String> getSubjectByLogin(String login) throws ExtSourceUnsupportedOperationException {
+		throw new ExtSourceUnsupportedOperationException();
+	}
+
+	@Override
+	public List<Map<String, String>> getGroupSubjects(Map<String, String> attributes) throws InternalErrorException {
+		//get pem file from url and parse it
+		String url = attributes.get(GroupsManager.GROUPMEMBERSQUERY_ATTRNAME);
+		PEMParser pemParser = parsePemFileFromURL(url);
+
+		//Prepare structure of all valid certificates mapped by login
+		Map<String, Pair<X509CertificateHolder, String>> validCertificatesForLogin = prepareStructureOfValidCertificates(pemParser);
+
+		List<Map<String, String>> subjects = new ArrayList<>();
+		//get subjects from map of valid certificates (every certificate is for 1 person in Perun)
+		for(String login: validCertificatesForLogin.keySet()) {
+			Map<String, String> subject = new HashMap<>();
+			subject.put("login", login);
+
+			//certificate need to be saved as map so we need to parse it correctly
+			Map<String, String> certificate = new LinkedHashMap<>();
+			certificate.put(validCertificatesForLogin.get(login).getLeft().getSubject().toString(), validCertificatesForLogin.get(login).getRight());
+			subject.put(attrUserCertificates, BeansUtils.attributeValueToString(certificate, LinkedHashMap.class.getName()));
+
+			//map on existing extSource with MU login
+			subject.put("additionalues_1", "https://idp2.ics.muni.cz/idp/shibboleth|cz.metacentrum.perun.core.impl.ExtSourceIdp|" + login + "@muni.cz|2");
+			subjects.add(subject);
+		}
+
+		return subjects;
+	}
+
+	//Private methods
+
+	/**
+	 * Create perunSession for ExtSourceTCS
+	 *
+	 * @return perun session for extSource TCS
+	 * @throws InternalErrorException if there is any problem to create perun session
+	 */
+	private PerunSession getSession() throws InternalErrorException {
+		final PerunPrincipal pp = new PerunPrincipal("ExtSourceTCS", ExtSourcesManager.EXTSOURCE_NAME_INTERNAL, ExtSourcesManager.EXTSOURCE_INTERNAL);
+		try {
+			return perunBl.getPerunSession(pp, new PerunClient());
+		} catch (InternalErrorException e) {
+			throw new InternalErrorException("Failed to get session for ExtSourceTCS.", e);
+		}
+	}
+
+	/**
+	 * Parse pem file from URL.
+	 *
+	 * @param url where the pem file should be find (generated by server)
+	 * @return PEMParser of pem file from the url
+	 * @throws InternalErrorException if there is any problem with parsing the file
+	 */
+	private PEMParser parsePemFileFromURL(String url) throws InternalErrorException {
+		try {
+			URL myURL = new URL(url);
+			//get pem file from url
+			HttpURLConnection con = (HttpURLConnection) myURL.openConnection();
+			con.setDoOutput(true);
+			con.setRequestProperty("Content-Type", "multipart/form-data;");
+			InputStream is = con.getInputStream();
+			return new PEMParser(new InputStreamReader(is));
```

Comment:
Also the InputStream. Shouldn't it be closed after we do not need it anymore?

Reply:
Right, add to close part.